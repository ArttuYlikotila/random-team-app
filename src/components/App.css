.App {
  text-align: center;
}

.wrapper {
  max-width: 80%;
  margin: auto;
}