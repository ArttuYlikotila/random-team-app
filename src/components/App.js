import React, { useState, useEffect } from 'react';
import './App.css';
import axios from 'axios';
import Spinner from './Spinner';
import Header from './Header';
import Controls from './controls/Controls';
import Tables from './tables/Tables';
import { makeTeams } from '../helpers/helperFunctions';

export default function App() {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(null);
  const [persons, setPersons] = useState([]);
  const [teamA, setTeamA] = useState([]);
  const [teamB, setTeamB] = useState([]);
  const [isTeamsReady, setIsTeamsReady] = useState(false);

  // Load 20 random person profiles from Random User API on first render
  useEffect(() => {
    setIsLoading(true);
    let mounted = true;

    axios.get('https://randomuser.me/api/', {params: {'results': 20}})
      .then(result => {
        if (mounted) {
          setPersons(result.data.results);
        }
      })
      .catch(error => {
        setError(error);
      })
      .then(() => {
        setIsLoading(false);
      })
    
    return () => {
      mounted = false;
    }
  }, []);

  // Build the teams and set the states of the app
  function toggleTeams() {
    const [firstTeam, secondTeam] = makeTeams(persons);

    setTeamA(firstTeam);
    setTeamB(secondTeam);

    setIsTeamsReady(true);
  }

  return (
    <div className="App" data-testid='App'>
      <Header />
      { error && <div>Error!</div> }
      { isLoading 
        ? <Spinner /> 
        : (!error
          ? <div className='wrapper'>
              <Controls toggleTeams={toggleTeams} isTeamsReady={isTeamsReady} setIsTeamsReady={setIsTeamsReady} teamA={teamA} />
              <Tables isTeamsReady={isTeamsReady} persons={persons} teamA={teamA} teamB={teamB} />
            </div>
          : null )
      }
    </div>
  );
}
