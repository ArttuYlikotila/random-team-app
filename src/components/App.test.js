import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import App from './App';

describe('App component', () => {
  test('Integration test of App component', async () => {
    render(<App />);

    expect(screen.getByTestId('App')).toMatchSnapshot();

    // Header is rendered
    expect(screen.getByRole('heading')).toBeInTheDocument();
    // Spinner is rendered
    expect(screen.getByRole('status')).toBeInTheDocument();
    // No buttons are rendered initially
    expect(screen.queryByRole('button')).toBeNull();

    // After users have been fetched, there is a table
    expect(await screen.findByRole('table')).toBeInTheDocument();
    // Button is rendered now as
    expect(screen.getByRole('button')).toBeInTheDocument();
    
    userEvent.click(screen.getByText('Create random teams'));

    // There are two tables now
    expect(screen.getAllByRole('table')).toHaveLength(2);
    // There are two buttons now
    expect(screen.getAllByRole('button')).toHaveLength(2);

    userEvent.click(screen.getByText('Show list of persons'));

    // There is only one table now
    expect(screen.getAllByRole('table')).toHaveLength(1);
  })
})
