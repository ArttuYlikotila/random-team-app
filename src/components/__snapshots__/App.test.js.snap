// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Integration test of App component 1`] = `
<div
  class="App"
  data-testid="App"
>
  <div
    class="pt-3 pb-3 mb-5 shadow"
  >
    <h1>
      Random Team App
    </h1>
  </div>
  <div
    class="text-center mt-4 w-100"
  >
    <div
      class="spinner-border"
      role="status"
    />
  </div>
</div>
`;
